core/runtime/src/main/java/io/quarkus/runtime/shutdown/ShutdownListener.java

```diff
@@ -0,0 +1,44 @@
+package io.quarkus.runtime.shutdown;
+
+/**
+ * A listener that can be registered to control the shutdown process and implement
+ * graceful shutdown.
+ *
+ * Shutdown happens in two phases. In the pre shutdown phase the application should
+ * function normally, but should notify external systems that it is about to go away.
+ *
+ * In the shutdown phase the app should disallow new external requests, however
+ * allow existing requests to complete normally.
+ */
+public interface ShutdownListener {
+
+    /**
+     * Pre shutdown notification, the listener can use this to notify external
+     * systems this application is about to shut down.
+     * 
+     * @param notification The notification event
+     */
+    default void preShutdown(ShutdownNotification notification) {
```

Comment:
in terms of API, any reason to not return a `CompletionStage`?

Reply:
Typically how are reported failures?

Reply:
Failures don't really matter, we are shutting down anyway, if a stage has failed then it needs to log and notify, so I am not sure if completion stage works here.